packages/web/src/common/store/cache/collections/addTrackToPlaylistSaga.ts

```diff
@@ -0,0 +1,288 @@
+  PlaylistOperations,
+  cacheActions,
+  getContext,
+  audioRewardsPageActions,
+  Collection,
+  Nullable,
+  SquareSizes,
+  UserTrackMetadata,
+  AudiusBackend,
+  UserCollection,
+  ID
+} from '@audius/common'
+import { isEqual } from 'lodash'
+import { call, put, select, takeEvery } from 'typed-redux-saga'
+
+import { make } from 'common/store/analytics/actions'
+import * as confirmerActions from 'common/store/confirmer/actions'
+import { confirmTransaction } from 'common/store/confirmer/sagas'
+import { ensureLoggedIn } from 'common/utils/ensureLoggedIn'
+import { waitForWrite } from 'utils/sagaHelpers'
+
+import { retrieveTracks } from '../tracks/utils'
+import { fetchUsers } from '../users/sagas'
+
+import { confirmOrderPlaylist } from './confirmOrderPlaylist'
+import {
+  retrieveCollection,
+  retrieveCollections
+} from './utils/retrieveCollections'
+
+const { getCollection } = cacheCollectionsSelectors
+const { setOptimisticChallengeCompleted } = audioRewardsPageActions
+
+type PlaylistWithArtwork = UserCollection & {
+  artwork: { file: File; url: string }
+}
+
+type AddTrackToPlaylistAction = ReturnType<
+  typeof cacheCollectionsActions.addTrackToPlaylist
+>
+
+async function getTrackArtworkUrl(
+  tracks: UserTrackMetadata[],
+  audiusBackend: AudiusBackend
+) {
+  return await Promise.all(
+    tracks.map(async (track) => {
+      const { user, cover_art_sizes, cover_art } = track
+      const userGateways = audiusBackend.getCreatorNodeIPFSGateways(
+        user.creator_node_endpoint
+      )
+
+      const imageUrl: string = await audiusBackend.getImageUrl(
+        cover_art_sizes || cover_art,
+        SquareSizes.SIZE_1000_BY_1000,
+        userGateways
+      )
+      return imageUrl
+    })
+  )
+}
+
+/** ADD TRACK TO PLAYLIST */
+
+export function* watchAddTrackToPlaylist() {
+  yield* takeEvery(
+    cacheCollectionsActions.ADD_TRACK_TO_PLAYLIST,
+    addTrackToPlaylistAsync
+  )
+}
+
+function* addTrackToPlaylistAsync(action: AddTrackToPlaylistAction) {
+  const { playlistId, trackId } = action
+  yield* waitForWrite()
+  const userId = yield* call(ensureLoggedIn)
+  const audiusBackendInstance = yield* getContext('audiusBackendInstance')
+  const web3 = yield* call(audiusBackendInstance.getWeb3)
+
+  // Retrieve tracks with the the collection so we confirm with the
+  // most up-to-date information.
+  const { collections } = yield* call(retrieveCollections, [playlistId], {
+    userId,
+    fetchTracks: true,
+    forceRetrieveFromSource: true
+  })
+
+  const playlist: Nullable<PlaylistWithArtwork> = collections[playlistId]
+  if (!playlist) return
+
+  // Retrieve track-to-add so we confirm with the
+  // most up-to-date information.
+  const [track] = yield* call(retrieveTracks, { trackIds: [trackId] })
+  const trackOwnerEntities = yield* call(fetchUsers, [track.owner_id])
+  const trackOwner = trackOwnerEntities.entries[track.owner_id]
+
+  if (track && trackOwner && playlist.track_count === 3) {
```

Comment:
do we want the track count be exactly 3 here?

Reply:
ah i see, i guess we would only need to generate the image again if the >4 tracks were reordered